### PR TITLE
Ensure TLS destructors run before thread joins in SGX

### PR DESCRIPTION
**Pull request by mzohreva**
The excellent test is from @jethrogb

For context see: https://github.com/rust-lang/rust/pull/83416#discussion_r617282907

### CODE REVIEW

File: library/std/src/sys/sgx/thread.rs [hunk 1]
```diff
@@ -3,32 +3,35 @@ use crate::ffi::CStr;
...
+            let (done, recv) = wait_notify::new();
             (Task { p, done }, recv)
         }
 
-        pub(super) fn run(self) {
+        pub(super) fn run(self, tls_guard: ActiveTls<'_>) {
             (self.p)();
-            let _ = self.done.send(());
+            drop(tls_guard);
+            let _ = self.done.notify();
```
Comment (jethrogb) [inline library/std/src/sys/sgx/thread.rs:34]: I'd prefer moving the notification out of here to entry() (as opposed to moving the TLS destructors in here)

Comment (jethrogb) [reply library/std/src/sys/sgx/thread.rs:34]: Perhaps with some kind of guard that notifies on drop

**Review by jethrogb — APPROVED** [review]
LGTM

**Review by dtolnay — APPROVED** [review]
Thanks -- accepting without particular scrutiny on the basis of @jethrogb's review.

**Comment by Dylan-DPC** [comment]
fails in [rollup](https://github.com/rust-lang/rust/pull/84509#issuecomment-826075097)

@bors r-

**Comment by mzohreva** [comment]
@dtolnay this is ready for another look.

@rustbot modify labels to +S-waiting-on-review, -S-waiting-on-author

**Comment by jethrogb** [comment]
Ping for review

**Comment by Dylan-DPC** [comment]
please dont ping for reviews, we normally give the reviewer 2 weeks to review it so they can pick it up when they are comfortable reviewing it.

**Comment by jethrogb** [comment]
I don't think it's reasonable to wait 2 weeks between every change.

**Comment by jethrogb** [comment]
bors r-

It's not clear from the test output, but it seems likely that the `join_orders_after_tls_destructors` test doesn't pass on Windows?

File: library/std/src/thread/local/tests.rs [hunk 2]
```diff
@@ -207,3 +208,109 @@ fn dtors_in_dtors_in_dtors_const_init() {
...
+
+                impl Drop for TlDrop {
+                    fn drop(&mut self) {
+                        loop {
+                            match SYNC_STATE.load(Ordering::SeqCst) {
+                                FRESH => thread::yield_now(),
+                                THREAD2_LAUNCHED => break,
+                                v => unreachable!("sync state: {}", v),
+                            }
+                        }
```
Comment (jethrogb) [inline library/std/src/thread/local/tests.rs:259]: After some debugging, it appears that on Windows, launching a thread and running TLS destructors in another thread are serialized. This loop should be moved to after line 279 `TL_DROP.with(|_| {})` to make sure thread 2 is launched before thread 1 enters TLS destructors.

**Comment by mzohreva** [comment]
@jethrogb I think you need @ with your bors command.

**Comment by apiraino** [comment]
Beta backport approved as per compiler team [on Zulip](https://zulip-archive.rust-lang.org/238009tcompilermeetings/21853weekly2021052054818.html#239597382)

@rustbot label +beta-accepted

File: library/std/src/thread/local/tests.rs [hunk 3]
```diff
@@ -207,3 +208,110 @@ fn dtors_in_dtors_in_dtors_const_init() {
...
+                Ordering::SeqCst,
+            ) {
+                Ok(_) => break,
+                Err(FRESH) => thread::yield_now(),
+                Err(THREAD2_LAUNCHED) => thread::yield_now(),
+                Err(THREAD2_JOINED) => {
+                    panic!("Main thread rendezvous after thread 2 joined thread 1")
+                }
+                v => unreachable!("sync state: {:?}", v),
+            }
```
Comment (tmiasko) [inline library/std/src/thread/local/tests.rs:313]: `compare_exchange_weak` allows spurious failures. If comparison succeeds but operation fails, the code will enter last unreachable branch and panic.

Comment (infinity0) [reply library/std/src/thread/local/tests.rs:313]: We observed this failure in #89584 on armhf Debian, I have started a re-try of the build to see if the error goes away. But yes it should be fixed so it never occurs.

Comment (jethrogb) [reply library/std/src/thread/local/tests.rs:313]: This was already fixed in https://github.com/rust-lang/rust/pull/86383 so you must've observed a different failure.

**Pull request merged**